File: services/worker_python/app/learning.py
import json

import httpx

from .schemas import ChatGptRepairRequest, LearningRequest, SocialAuditAnalysis, StudyAnalysis
from .settings import settings


async def _structured(prompt: str, schema_model):
    if not settings.ollama_url:
        raise RuntimeError("Ollama is not configured for the transcription worker.")
    schema = schema_model.model_json_schema()
    async with httpx.AsyncClient(timeout=1200) as client:
        response = await client.post(
            f"{settings.ollama_url.rstrip('/')}/api/chat",
            json={
                "model": settings.ollama_model,
                "messages": [{"role": "user", "content": prompt}],
                "format": schema,
                "stream": False,
                "think": False,
                "keep_alive": "10m",
                "options": {"temperature": 0.1},
            },
        )
        response.raise_for_status()
        payload = response.json()
        content = payload.get("message", {}).get("content", "").strip()
        if not content:
            raise RuntimeError(f"Ollama returned no structured output (reason: {payload.get('done_reason', 'unknown')}).")
        return schema_model.model_validate_json(content)


def _study_targets(transcript: str) -> tuple[int, int, int]:
    words = len(transcript.split())
    if words < 600:
        return 6, 4, 5
    if words < 1800:
        return 10, 6, 8
    if words < 5000:
        return 14, 8, 10
    return 18, 10, 12


def _study_needs_expansion(pack: StudyAnalysis, transcript: str) -> bool:
    flashcards, quiz, test = _study_targets(transcript)
    if not pack.topics:
        return True
    if len(pack.flashcards) < flashcards or len(pack.quiz) < quiz or len(pack.test) < test:
        return True
    if len(transcript.split()) >= 800 and any(len(topic.summary.split()) < 70 for topic in pack.topics):
        return True
    return False


async def _expand_study_pack(request: LearningRequest, draft: StudyAnalysis) -> StudyAnalysis:
    schema = StudyAnalysis.model_json_schema()
    flashcards, quiz, test = _study_targets(request.transcript)
    prompt = f"""Audit and expand this draft study pack into a complete self-contained course based ONLY on the source transcript.

The learner's requirement is strict: they should be able to learn the material from the generated pack without going back to the video or transcript for ordinary study.

Rules:
- Preserve useful content and stable IDs from the draft whenever possible.
- Stay grounded in the transcript. Do not add outside facts merely to hit a count.
- Every substantive transcript concept needs a topic.
- Each topic's `summary` is the actual LESSON, not a teaser. Teach definitions, reasoning, relationships, syntax, process, edge cases, and context that the transcript explains. For a normal topic, write several useful paragraphs or an equivalently detailed explanation. Use line breaks when it improves readability.
- `keyPoints` should capture the rules and facts the learner must remember.
- `examples` should contain concrete worked examples, code, commands, scenarios, or applications when the transcript supports them.
- `commonMistakes` should explain likely misunderstandings and why they are wrong.
- `prerequisites` should identify earlier topics that should be understood first.
- Flashcards must cover the important definitions, distinctions, rules, syntax, edge cases, and common mistakes. Target AT LEAST {flashcards} useful cards when the transcript supports that much material.
- Quiz must be a real checkpoint after learning. Target AT LEAST {quiz} questions with broad topic coverage.
- Test must be a real final assessment, harder and broader than the quiz. Target AT LEAST {test} questions, mixing recall, explanation, application, edge cases, and code/reasoning where appropriate.
- Do not make the test a single-question self-check.
- Practice tasks come AFTER learning and assessment. They should be hands-on exercises that prove the learner can apply the material. They are not substitutes for teaching notes.
- Every question needs a correct answer and a teaching explanation.
- Multiple-choice questions need plausible distractors and one best answer. Other question types use an empty choices list.
- Keep question type to exactly one of: multiple_choice, short_answer, explain, code.
- reviewScheduleDays should normally be [1,3,7,14,30].
- Return only JSON matching this schema: {json.dumps(schema, ensure_ascii=False)}

Requested title: {request.title}
Context supplied by learner: {request.context}

DRAFT PACK:
{draft.model_dump_json(indent=2)}

SOURCE TRANSCRIPT:
{request.transcript}
"""
    return await _structured(prompt, StudyAnalysis)


async def create_study_pack(request: LearningRequest) -> StudyAnalysis:
    schema = StudyAnalysis.model_json_schema()
    flashcards, quiz, test = _study_targets(request.transcript)
    prompt = f"""Build a complete self-contained course from this lecture/course transcript.

The learner wants to study from THIS generated pack instead of rewatching the source. This is not a summary task. Teach the material first, then create active recall and assessment.

Rules:
- Stay grounded in the source transcript. Do not invent course facts or silently add outside knowledge.
- Correct obvious speech-to-text noise only when the intended meaning is clear.
- Create a useful nested topic hierarchy. `suggestedPath` should be a short library path such as ["Java", "Classes", "Inheritance"].
- Topics must use stable short IDs and parentId to express hierarchy. Root topics use parentId=null.
- Each topic's `summary` must function as a detailed lesson the learner can study directly. Explain what the concept is, how it works, why it matters in the transcript, important distinctions, relationships to nearby concepts, and edge cases the transcript covers. Do not use one-sentence summaries for substantive topics.
- Put memorable rules and facts in `keyPoints`.
- Put concrete examples, code, commands, scenarios, or worked applications in `examples` whenever supported by the transcript.
- Put likely misconceptions and their correction in `commonMistakes`.
- Learning objectives must describe things the learner should be able to DO or EXPLAIN after studying.
- Flashcards are mandatory when the transcript contains studyable facts. Target AT LEAST {flashcards} useful cards when supported, distributed across the substantive topics.
- Quiz is mandatory when there is enough material. Target AT LEAST {quiz} questions and use it as a broad comprehension checkpoint after learning.
- Test is mandatory when there is enough material. Target AT LEAST {test} questions and make it clearly harder than the quiz, covering explanation, application, edge cases, and code/reasoning when appropriate.
- Never collapse a normal lesson into one quiz/test question.
- Multiple-choice questions must have plausible distractors and exactly one best answer.
- For short-answer/explain/code questions, choices must be an empty list.
- Every question needs the correct answer and a teaching explanation.
- Practice tasks happen after learning. They must require observable work and include explicit success criteria. Do not use tasks as a replacement for the teaching material.
- Include enough material to cover every substantive topic in the transcript without bloating trivial points.
- reviewScheduleDays should normally be [1,3,7,14,30].
- Output only JSON matching this schema: {json.dumps(schema, ensure_ascii=False)}

Requested title: {request.title}
Context supplied by learner: {request.context}

SOURCE TRANSCRIPT:
{request.transcript}
"""
    pack = await _structured(prompt, StudyAnalysis)
    if _study_needs_expansion(pack, request.transcript):
        pack = await _expand_study_pack(request, pack)
    return pack


async def create_social_audit(request: LearningRequest) -> SocialAuditAnalysis:
    schema = SocialAuditAnalysis.model_json_schema()
    prompt = f"""Analyze this conversation as an evidence-based communication coach.

The purpose is to help the learner notice habits, reasoning patterns, emotional communication, and social norms they may not naturally notice. This is NOT therapy, diagnosis, lie detection, or mind-reading.

Rules:
- Separate observable evidence from interpretation.
- Never diagnose a mental health condition, personality disorder, neurotype, attachment style, or hidden motive.
- Never state another person's intent, emotion, or belief as fact unless they explicitly said it.
- For every non-obvious interpretation, include plausible alternative explanations.
- Confidence must reflect how strongly the transcript supports the observation. Use lower confidence when context is missing.
- Emotional observations may cover tone implied by words, validation, escalation/de-escalation, emotional bids, defensiveness, reassurance, and unmet communication needs, but only as possibilities supported by text.
- Logical observations may cover assumptions, contradictions, missing evidence, overgeneralization, strong reasoning, question quality, and whether conclusions follow from what was said.
- Social observations may cover turn-taking, reciprocity, boundaries, politeness, directness, status/power cues, interruptions visible in transcript structure, topic shifts, repair attempts, and expectations that are common social norms.
- habitsAndPatterns should identify repeated communication behaviors visible in THIS conversation. Do not claim a lifelong pattern from one sample.
- possibleBlindSpots should be framed as things worth checking, not verdicts.
- strengths matter. Include things the learner did effectively.
- socialNormsWorthLearning should explain concrete social conventions or expectations that would help in similar situations, with examples.
- lessons should be practical and specific.
- experiments should be small behaviors the learner can try in future conversations and observe the result.
- uncertaintyNotes must explicitly list important missing context that could change the interpretation.
- Do not moralize ordinary differences in communication style.
- If the transcript contains conflict, present reasonable interpretations of all sides instead of automatically siding with the learner.
- Output only JSON matching this schema: {json.dumps(schema, ensure_ascii=False)}

Conversation title: {request.title}
Context supplied by learner: {request.context}

SOURCE TRANSCRIPT:
{request.transcript}
"""
    return await _structured(prompt, SocialAuditAnalysis)


async def repair_study_chatgpt_output(request: ChatGptRepairRequest) -> StudyAnalysis:
    schema = StudyAnalysis.model_json_schema()
    prompt = f"""Convert the pasted ChatGPT study response below into the application's exact StudyAnalysis schema.

This is a STRUCTURAL REPAIR task, not a new study-generation task.

Rules:
- Preserve the user's useful study content, wording, IDs, topic hierarchy, answers, examples, flashcards, quiz questions, test questions, tasks, and review schedule whenever possible.
- Repair malformed JSON, Markdown escaping, broken quotes, code formatting, and wrapper text instead of discarding content.
- Remove Markdown escape backslashes that only exist because the response was copied from rendered chat.
- Preserve existing IDs exactly whenever they are present and unique.
- Do not invent new subject-matter claims just to fill fields.
- Every question type MUST become one of: multiple_choice, short_answer, explain, code.
- Map variants such as multiple-choice or multiple\\_choice to multiple_choice.
- Map code_writing, code_reasoning, code_review, coding, implementation, or similar programming tasks to code.
- Map scenario, reasoning, analysis, discussion, or explanation-style questions to explain unless short_answer is clearly more appropriate.
- Non-multiple-choice questions must use an empty choices array.
- If a task contains `instructions` as an array, combine those steps into one readable `instruction` string without losing steps.
- If a task contains `successCriteria` as an array, combine the criteria into one readable `successCriteria` string without losing criteria.
- Keep difficulty limited to easy, medium, or hard. Use medium only when the source does not make the level clear.
- Ensure all required strings exist and all required arrays are arrays.
- Keep reviewScheduleDays as positive integers.
- Return only JSON matching this schema: {json.dumps(schema, ensure_ascii=False)}

Fallback title if the response has none: {request.title}

PASTED CHATGPT RESPONSE:
{request.response}
"""
    return await _structured(prompt, StudyAnalysis)


async def repair_audit_chatgpt_output(request: ChatGptRepairRequest) -> SocialAuditAnalysis:
    schema = SocialAuditAnalysis.model_json_schema()
    prompt = f"""Convert the pasted ChatGPT conversation-audit response below into the application's exact SocialAuditAnalysis schema.

This is a STRUCTURAL REPAIR task, not a new psychological or conversational analysis.

Rules:
- Preserve the useful observations and wording from the pasted response whenever possible.
- Repair malformed JSON, Markdown escaping, wrapper prose, missing field names, and array-vs-string mistakes instead of discarding content.
- Do not diagnose anyone and do not add claims about hidden intent.
- emotional, logical, social, habitsAndPatterns, and possibleBlindSpots must contain observation objects with title, evidence, interpretation, alternatives, and confidence.
- Turn a single evidence string into a one-item array. Do the same for alternatives.
- If an observation title is missing, derive a short neutral title from its interpretation without adding a new claim.
- Confidence must be numeric from 0 to 1. Convert percentages and high/medium/low wording when necessary.
- strengths, lessons, reflectionQuestions, and uncertaintyNotes must be arrays of strings.
- socialNormsWorthLearning items must contain norm, whyItMatters, and example.
- experiments items must contain title, action, and whatToNotice.
- Do not omit useful sections merely because the pasted structure used different field names.
- Return only JSON matching this schema: {json.dumps(schema, ensure_ascii=False)}

Fallback title if the response has none: {request.title}

PASTED CHATGPT RESPONSE:
{request.response}
"""
    return await _structured(prompt, SocialAuditAnalysis)

File: src/services/learning.ts
import { z } from 'zod';
import type { SocialAuditAnalysis, StudyAnalysis } from '../domain/learning';
import type { LanguageMode } from '../domain/types';

const difficulty = z.enum(['easy', 'medium', 'hard']);
const topicSchema = z.object({
  id: z.string(),
  title: z.string(),
  parentId: z.string().nullable().optional(),
  summary: z.string(),
  keyPoints: z.array(z.string()).default([]),
  examples: z.array(z.string()).default([]),
  commonMistakes: z.array(z.string()).default([]),
  prerequisites: z.array(z.string()).default([])
});
const questionSchema = z.object({
  id: z.string(),
  topicId: z.string(),
  type: z.enum(['multiple_choice', 'short_answer', 'explain', 'code']),
  prompt: z.string(),
  choices: z.array(z.string()).default([]),
  answer: z.string(),
  explanation: z.string(),
  difficulty
});
const studySchema = z.object({
  title: z.string(),
  suggestedPath: z.array(z.string()).default([]),
  overview: z.string(),
  learningObjectives: z.array(z.string()).default([]),
  topics: z.array(topicSchema).default([]),
  flashcards: z.array(z.object({
    id: z.string(), topicId: z.string(), front: z.string(), back: z.string(), difficulty
  })).default([]),
  quiz: z.array(questionSchema).default([]),
  test: z.array(questionSchema).default([]),
  tasks: z.array(z.object({
    id: z.string(), topicId: z.string(), title: z.string(), instruction: z.string(),
    successCriteria: z.string(), difficulty
  })).default([]),
  reviewScheduleDays: z.array(z.number().int().positive()).default([1, 3, 7, 14, 30])
});

const observationSchema = z.object({
  title: z.string(),
  evidence: z.array(z.string()).default([]),
  interpretation: z.string(),
  alternatives: z.array(z.string()).default([]),
  confidence: z.number().min(0).max(1)
});
const socialAuditSchema = z.object({
  title: z.string(),
  summary: z.string(),
  emotional: z.array(observationSchema).default([]),
  logical: z.array(observationSchema).default([]),
  social: z.array(observationSchema).default([]),
  habitsAndPatterns: z.array(observationSchema).default([]),
  possibleBlindSpots: z.array(observationSchema).default([]),
  strengths: z.array(z.string()).default([]),
  socialNormsWorthLearning: z.array(z.object({ norm: z.string(), whyItMatters: z.string(), example: z.string() })).default([]),
  lessons: z.array(z.string()).default([]),
  reflectionQuestions: z.array(z.string()).default([]),
  experiments: z.array(z.object({ title: z.string(), action: z.string(), whatToNotice: z.string() })).default([]),
  uncertaintyNotes: z.array(z.string()).default([])
});

const youtubeImportSchema = z.object({
  title: z.string(),
  sourceName: z.string(),
  transcript: z.string().min(1),
  method: z.enum(['captions', 'whisper']),
  videoId: z.string(),
  webpageUrl: z.string(),
  durationSeconds: z.number().int().nonnegative().nullable().optional()
});

async function postJson<T>(workerUrl: string, path: string, body: unknown, schema: z.ZodType<T>): Promise<T> {
  const response = await fetch(`${workerUrl.replace(/\/$/, '')}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  });
  if (!response.ok) {
    const raw = await response.text();
    let detail = raw;
    try {
      const parsed = JSON.parse(raw) as { detail?: string };
      detail = parsed.detail || raw;
    } catch {
      // Keep the raw response when the worker did not return JSON.
    }
    throw new Error(detail || `Learning worker returned ${response.status}`);
  }
  return schema.parse(await response.json());
}

function parseChatGptJson(value: string): unknown {
  const trimmed = value.trim();
  if (!trimmed) throw new Error('Paste the ChatGPT result first.');

  try {
    return JSON.parse(trimmed);
  } catch {
    // ChatGPT may wrap the replacement object in a JSON code fence.
  }

  const fenced = [...trimmed.matchAll(/```json\s*([\s\S]*?)```/gi)];
  for (const match of fenced.reverse()) {
    try {
      return JSON.parse(match[1].trim());
    } catch {
      // Try another fenced block or the broad object fallback below.
    }
  }

  const start = trimmed.indexOf('{');
  const end = trimmed.lastIndexOf('}');
  if (start >= 0 && end > start) {
    try {
      return JSON.parse(trimmed.slice(start, end + 1));
    } catch {
      // Fall through to the clearer error below.
    }
  }
  throw new Error('I could not find a valid JSON study/audit result in that ChatGPT response. Ask ChatGPT to return the complete replacement JSON object and paste it again.');
}

export function analyzeStudyWithOllama(workerUrl: string, input: { title: string; transcript: string; context: string }): Promise<StudyAnalysis> {
  return postJson(workerUrl, '/v1/study', input, studySchema);
}

export function analyzeSocialAuditWithOllama(workerUrl: string, input: { title: string; transcript: string; context: string }): Promise<SocialAuditAnalysis> {
  return postJson(workerUrl, '/v1/social-audit', input, socialAuditSchema);
}

export function importYouTubeTranscript(workerUrl: string, input: { url: string; languageMode: LanguageMode; context: string }) {
  return postJson(workerUrl, '/v1/youtube/transcript', {
    url: input.url,
    language_mode: input.languageMode,
    context: input.context
  }, youtubeImportSchema);
}

export function parseStudyChatGptResult(value: string): StudyAnalysis {
  return studySchema.parse(parseChatGptJson(value));
}

export function parseSocialAuditChatGptResult(value: string): SocialAuditAnalysis {
  return socialAuditSchema.parse(parseChatGptJson(value));
}

export function buildStudyChatGptPrompt(title: string, transcript: string, analysis: StudyAnalysis): string {
  return `You are my expert tutor and curriculum designer. Improve this locally generated study pack without deleting useful material. Correct mistakes against the transcript, improve the hierarchy, add missing active-recall questions, make the test genuinely challenging, and make the practice tasks prove mastery rather than passive reading. Keep claims grounded in the source.

IMPORTANT RETURN CONTRACT:
- Return the COMPLETE replacement study pack as one valid JSON object only.
- Keep exactly the same top-level structure and field names as LOCAL OLLAMA STUDY PACK.
- Preserve existing ids when an item still represents the same topic/question/task/flashcard so my learning progress can be preserved.
- New items must get short unique string ids.
- Do not add commentary before or after the JSON.
- Do not omit arrays just because they are empty.

TITLE:
${title}

LOCAL OLLAMA STUDY PACK:
${JSON.stringify(analysis, null, 2)}

SOURCE TRANSCRIPT:
${transcript}`;
}

export function buildAuditChatGptPrompt(title: string, transcript: string, analysis: SocialAuditAnalysis): string {
  return `Review this conversation audit as a careful coach. Improve the emotional, logical, and social analysis while staying evidence-based. Do not diagnose anyone, do not claim hidden intent as fact, distinguish observation from interpretation, add plausible alternative explanations, and focus on habits, social norms, communication choices, and practical lessons I can test.

IMPORTANT RETURN CONTRACT:
- Return the COMPLETE replacement audit as one valid JSON object only.
- Keep exactly the same top-level structure and field names as LOCAL OLLAMA AUDIT.
- Keep confidence values between 0 and 1.
- Do not add commentary before or after the JSON.
- Do not omit arrays just because they are empty.

TITLE:
${title}

LOCAL OLLAMA AUDIT:
${JSON.stringify(analysis, null, 2)}

SOURCE TRANSCRIPT:
${transcript}`;
}

export async function copyAndOpenChatGpt(prompt: string): Promise<void> {
  await navigator.clipboard.writeText(prompt);
  window.open('https://chatgpt.com/', '_blank', 'noopener,noreferrer');
}
